packages/material-ui/src/Button/Button.js

```diff
@@ -278,6 +279,22 @@ const Button = React.forwardRef(function Button(props, ref) {
     ...other
   } = props;
 
+  const themeVariantsClasses = useThemeVariants(
+    {
+      color,
+      component,
+      disabled,
+      disableElevation,
+      disableFocusRipple,
+      fullWidth,
+      size,
+      type,
+      variant,
+      ...props,
```

Comment:
Question about spreading props:
If I pass `{ variant: 'undefined' }` then `variant` will default to `'primary'`. But then we later "spread over it". Does it ignore explicit undefined i.e. is `{ variant: 'primary', ...{ variant: undefined }}` actually `{ variant: 'primary' }` or `{ variant: undefined }` (which is my intuition).

Edit: I can just use chrome devtools. It actually produces `{ variant: undefined }`:
![chrome devtools output for `{ variant: 'primary', ...{ variant: undefined }}` which is `{ variant: undefined }`](https://user-images.githubusercontent.com/12292047/88526677-84815c80-cffc-11ea-8fec-02932c0980fe.png)
 
Shouldn't we 
```suggestion
      ...other,
```
?

Reply:
Just realized that this isn't correct either. Now we're missing the destructured props that don't have a default value.

`useThemeVariants({ ...props, disabled })` should be correct. Now we guarantee existence of all props and that default values are included.

Reply:
I still feel like:
```
{      
  size,
  type,
  variant,
  // other default values
  ...props    
}
``` 
is correct as if you explicitly set `variant: undefined` it would win, and I believe that is the expected...

I don't quite understand how 
```
useThemeVariants({ ...props, disabled })
```

is correct, where are the defaults specified when destructing?

Reply:
> is correct as if you explicitly set variant: undefined it would win, and I believe that is the expected...

It wouldn't inside the component though. `<Button variant={undefined} />` would still render the `primary` variant but your matcher would never match it.

> I don't quite understand how `useThemeVariants({ ...props, disabled })` is correct, where are the defaults specified when destructing?

We're assigning here. Destructuring happens in `const { disabled = false, ...other } = props`;

Reply:
i.e. why would you want `<Button variant={undefined} />` to render a primary Button but not have your `{ variant: 'primary' }` matcher applied? `<Button variant={undefined} />`, `<Button varian="primary" />` and `<Button />` would all render the same thing but `{ variant: 'primary' }` would only apply to two. Since we never distinguish in React between a prop being `undefined` because it's not declarated and it being `undefined` because it is declared as such we should have a strong reason for considering this difference in this API.

Reply:
I udnerstand, in that case shuold we just specify all props which are destructured (the ones with defaults and the ones without), and than spread other?

```useThemeVariants({ color, variant, disabled, ...other })```

Reply:
> useThemeVariants({ color, variant, disabled, ...other })

This was my initial suggestion but this requires to specify all destructured props (e.g. renamed props).

`{...props, disabled}` is the shortest one (all props + default values are preferred). Why does that not work for you?